### ISSUE TITLE
Stack overflow while checking mathlib

### ISSUE THREAD

**Issue by Kha** [ammkrn/nanoda#3]
I exported all of mathlib like this:
```
$ git rev-parse HEAD
cc04ba7e5641d6c038d93d83dc163e2ae0d09709
$ lean --recursive --export=mathlib.txt src
$ du -h mathlib.txt
191M	mathlib.txt
```
and then tried to check it:
```
$ ./target/release/nanoda mathlib.txt

thread '<unknown>' has overflowed its stack
fatal runtime error: stack overflow
[1]    12335 abort      ./target/release/nanoda mathlib.txt
```
Am I doing something wrong? I'm on Linux; I tried to increase the stack size with `ulimit -s unlimited`, but I guess that doesn't help because it only affects the main thread.

`gdb` says that the overflow happens after more than 6000 stack frames of defeq checking. I don't know of which declaration.

**Comment by ammkrn** [comment]
No, this is definitely my fault. The troublemaker is `pi_gt_314` from `src/data/real/pi.lean`; the mathlib export file I was using during testing was one I made when I first started working on this which didn't include that definition (it looks like it was added in April). If I skip that definition in particular, everything else type checks since it isn't used anywhere else, I just didn't properly prepare for terms that large. I did make a concerted effort to use iteration instead of recursion where possible specifically due to concerns about stack growth but evidently that wasn't enough, so I'll put the recursor transition aside for the moment and get to work on this.

Thanks for bringing this to my attention.

**Comment by ammkrn** [comment]
It should be fixed for now. The overflow was coming from the (Const, Const) branch of the definitional equality check where you check all of the subtrees of the unfolded lhs/rhs terms. It's only an issue with that one declaration (so far) and the explicit stack growth to accommodate it seems to work fine, but I'll keep looking for a better long term solution that prevents it in the first place. 
Let me know if there are any issues on your system.

Also, in testing with the stacker crate, I've found that if I build and run the binary in one step, on the very first run the program will sometimes terminate (after checking everything) with a segfault, but every subsequent execution runs fine, so you might experience that. I don't know what causes it.

**Comment by Kha** [comment]
Oh, that is an interesting crate! I wonder how the proof passed `leanchecker` (and trepplein); either your reduction heuristics are a bit different, your stack frames are larger, or the slightly raised stack size of Lean threads (8MB) is just enough to make the proof pass. I guess if Lean overflowed its stack, the proof would never have been comitted, so it must fit into its stack limit by definition. So perhaps raising the limit could be good enough for your checker as well.

**Comment by ammkrn** [comment]
I think it's a little bit of both; there are other declarations that require between 2 and 4 MB stack space to check in nanoda, but that one declaration grows past 8MB as things are now, so there are definitely some differences in implementation that I haven't found yet. The part of the definitional equality check I thought might be different is the same between this, Lean, and Trepplein though; two Const items with the same name and levels have their unfolded applications recursively checked. The only difference seems to be that nanoda and trepplein check smallest -> largest subtree, but Lean checks largest -> smallest since it hasn't already despined the tree into a buffer.
Aside from the stack growth, I'm not happy with how much memory this consumes when checking parts of the newer mathlib, so if I can't uncover the secrets of the C++ version while doing the recursor stuff I'll work on something specific to this implementation.

**Comment by ammkrn** [comment]
See reduction rules issue. Will revisit this in December.

**Comment by ammkrn** [comment]
This is semi-officially resolved in 0.2.0 with the removal of reduction rules. I haven't re-implemented threads, but when I do I'll make sure the smaller spawned thread stacks don't overflow.
